File: src/com/train/algorithm/twoPointers/IntervalListIntersections.cpp
//
// Created by 罗斯 on 27/8/2022.
//
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        size_t i = 0;
        size_t j = 0;
        vector<vector<int>> res;
        while(i < firstList.size() && j < secondList.size()){
            const int s = max(firstList[i][0],secondList[j][0]);
            const int e = min(firstList[i][1],secondList[j][1]);
            if(s <= e) res.push_back({s,e});
            if(firstList[i][1] < secondList[j][1])
                ++i;
            else
                ++j;
        }
        return res;
    }
};

File: src/com/train/algorithm/twoPointers/implementInPython/IntervalListIntersections.py
class Solution(object):
    def intervalIntersection(self, firstList, secondList):
        i,j = 0,0
        res = []
        while i < len(firstList) and j < len(secondList):
            startA,endA = firstList[i]
            startB,endB = secondList[j]
            if startA <= endB  and startB <= endA:
                res.append([max(startB,startA),min(endB,endA)])
            if endA <= endB:
                i += 1
            else:
                j += 1
        return res;